### PR TITLE
add non static call constraint for sstore

### PR DESCRIPTION
**Pull request by DreamWuGit**
sstore op code will modify state, so if it is executed in static call , exception will occur ,  here constrain only in non static call , spec https://github.com/appliedzkp/zkevm-specs/pull/188

### CODE REVIEW

**Comment by DreamWuGit** [comment]
the CI error looks weird to me (https://github.com/appliedzkp/zkevm-circuits/runs/6073405191?check_suite_focus=true), Does anybody know that ? @CPerezz  @ChihChengLiang  @miha-stopar ?

**Comment by ChihChengLiang** [comment]
@DreamWuGit Created an issue and I'm investigating. For reviewers, Feel free to ignore that error and do the review.

File: bus-mapping/src/evm/opcodes/sstore.rs [hunk 1]
```diff
@@ -33,6 +33,15 @@ impl Opcode for Sstore {
                 value: Word::from(state.tx_ctx.id()),
             },
         );
+        state.push_op(
```
Comment (roynalnaruto) [inline bus-mapping/src/evm/opcodes/sstore.rs:36]: Looks good overall!

I think we should also assert the `CallContext` lookups being added in this bus-mapping implementation, along with the `StackOp` and `StorageOp` checks.

File: bus-mapping/src/evm/opcodes/sstore.rs [hunk 2]
```diff
@@ -187,7 +196,7 @@ mod sstore_tests {
             .unwrap();
 
         assert_eq!(
```
Comment (roynalnaruto) [inline bus-mapping/src/evm/opcodes/sstore.rs:198]: Before this we could check that the `CallContext` ops were pushed in the appropriate/expected order?

**Review by roynalnaruto — COMMENTED** [review]
Minor suggestions from my side, looks good otherwise :)

File: bus-mapping/src/evm/opcodes/sstore.rs (hunk 1 above)
Comment (DreamWuGit) [reply bus-mapping/src/evm/opcodes/sstore.rs:36]: you mean assert `CallContext` ops in below unit test  ?

**Comment by AronisAt79** [comment]
> the CI error looks weird to me (https://github.com/appliedzkp/zkevm-circuits/runs/6073405191?check_suite_focus=true), Does anybody know that ? @CPerezz @ChihChengLiang @miha-stopar ?

@DreamWuGit this happens because github.secrets are not shared with workflows triggered from PRs on forks.Therefore the invoked action is not permitted to access the   https://api.github.com/repos/appliedzkp/zlevm-circuits/pull/467/requested_reviewers endpoint. We are working on alternatives with @ntampakas

File: bus-mapping/src/evm/opcodes/sstore.rs (hunk 1 above)
Comment (roynalnaruto) [reply bus-mapping/src/evm/opcodes/sstore.rs:36]: yes

**Comment by ntampakas** [comment]
We are trying to avoid GH Actions limitations on forked repos, by implementing two different workflows:
1. An action, triggered by the PR coming from the fork, that will create and upload an artifact that will contain all the necessary info (f.x.: PR_NUM, REPO, etc.)
2. A cron job that will be executed periodically on the base repo, after downloading the artifact mentioned above and assigning randomly selected reviewers to the PR in question.

This is the cleanest way to trigger workflows from PRs coming from forks, that bypasses the necessity to access the GH secret(s). 
There are a couple of drawbacks:
Github does no guarantee that the workflow using the schedule event, will run at the pre-defined time.

**Comment by ntampakas** [comment]
Following up on the previous message, here is the implementation we have concluded on with pros and cons.

1. Every PR (either from fork or not) creates and uploads an artifact on base repo, consisting of all PR info.
2. A second workflow runs periodically as a scheduled task, downloads and extracts the artifact and in general may run any action requiring access to GitHub token: In our case the action is to randomly select reviewer(s) and assign them in PR.

Pros:
No need to distribute tokens to external accounts, as first artifact creation requires no write permission.
Same concept can be used for providing benchmark results in comments.

Cons:
Second workflow is executed as a cron job, so it will flood the actions history with workflow run records. It will make it a bit annoying to read the actions history (without filtering).
As mentioned in the comment above, Github does no guarantee that the workflow using the schedule event, will run at the pre-defined time. Therefore reviewers assignment, might take more than expected.

If everyone agrees on this approach, i will shortly raise a PR for review.
@ed255 @CPerezz @ChihChengLiang

**Review by icemelon — APPROVED** [review]
overall LGTM. Just need to address the comments

**Comment by DreamWuGit** [comment]
@ntampakas  I am curious why other PR not hit this issue ， even my other PR (i.e. #466  are OK) :）

File: bus-mapping/src/evm/opcodes/sstore.rs (hunk 2 above)
Comment (DreamWuGit) [reply bus-mapping/src/evm/opcodes/sstore.rs:198]: updated in https://github.com/appliedzkp/zkevm-circuits/pull/467/commits/7e7470ad9b0a7a602e0348d70c81c65e59a57cc2 thanks

File: bus-mapping/src/evm/opcodes/sstore.rs (hunk 1 above)
Comment (DreamWuGit) [reply bus-mapping/src/evm/opcodes/sstore.rs:36]: updated in https://github.com/appliedzkp/zkevm-circuits/commit/7e7470ad9b0a7a602e0348d70c81c65e59a57cc2 thanks

**Comment by DreamWuGit** [comment]
> other



> @ntampakas I am curious why other PR not hit this issue ， even my other PR (i.e. #466 is OK) :）

after push new stuff,  CI can pass now .

**Comment by CPerezz** [comment]
> Following up on the previous message, here is the implementation we have concluded on with pros and cons.
> 
>     1. Every PR (either from fork or not) creates and uploads an artifact on base repo, consisting of all PR info.
> 
>     2. A second workflow runs periodically as a scheduled task, downloads and extracts the artifact and in general may run any action requiring access to GitHub token: In our case the action is to randomly select reviewer(s) and assign them in PR.
> 
> 
> Pros: No need to distribute tokens to external accounts, as first artifact creation requires no write permission. Same concept can be used for providing benchmark results in comments.
> 
> Cons: Second workflow is executed as a cron job, so it will flood the actions history with workflow run records. It will make it a bit annoying to read the actions history (without filtering). As mentioned in the comment above, Github does no guarantee that the workflow using the schedule event, will run at the pre-defined time. Therefore reviewers assignment, might take more than expected.
> 
> If everyone agrees on this approach, i will shortly raise a PR for review. @ed255 @CPerezz @ChihChengLiang

@AronisAt79 @ntampakas could you move this thread to the issue/PR where it corresponds to?

**Review by CPerezz — APPROVED** [review]
LGTM.

Can we update and merge this??

**Comment by CPerezz** [comment]
@lispc I can't rebase myself in order to merge it. 
Please do so and ping me on TG or here in the conversation once the rebase has passed the CI so that I can merge this PR :smile:

**Comment by lispc** [comment]
i don't fully understand "rebase myself in order to merge it".

I clicked the "update branch" button. And i think there is a "squash and merge" button, which is very helpful for things like this?

image

**Comment by CPerezz** [comment]
Oh @lispc I thought you couldn't merge the PR.

Ye, just squash and merge please! :)

**Comment by lispc** [comment]
nonono, I cannot merge this.  I snapshot from another repo just to give a demo..

**Pull request merged**